store/sqlstore/post_store.go

```diff
@@ -1202,3 +1227,57 @@ func (s SqlPostStore) GetOldest() store.StoreChannel {
 		result.Data = &post
 	})
 }
+
+// GetMaxPostSize returns the maximum number of runes that may be stored in a post.
+func (s SqlPostStore) GetMaxPostSize(allowFromCache bool) store.StoreChannel {
+	return store.Do(func(result *store.StoreResult) {
+		var maxPostSize int32 = model.POST_MESSAGE_MAX_RUNES_V1
+
+		if allowFromCache && maxPostSizeCached != nil {
+			result.Data = *maxPostSizeCached
+			return
+		}
+
+		if s.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+			// PostgreSQL uses TEXT as the underlying type, and we will have safely
+			// migrated from any VARCHAR(4000) with the same underlying type on
+			// upgrade.
+			maxPostSize = model.POST_MESSAGE_MAX_RUNES_V2
+
+		} else if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+			// Unlike PostgreSQL, MySQL's VARCHAR and TEXT types do not share the same
+			// underlying storage strategy. Honour any existing limitation, and rely
+			// on system administrators to migrate the column if necessary.
+			var maxPostSizeBytes int32
+
+			if err := s.GetReplica().SelectOne(&maxPostSizeBytes, `
+				SELECT 
+					CHARACTER_MAXIMUM_LENGTH
+				FROM 
+					INFORMATION_SCHEMA.COLUMNS
+				WHERE 
+					table_schema = DATABASE()
+				AND	table_name = 'Posts'
+				AND	column_name = 'Message'
+				LIMIT 0, 1
+			`); err != nil {
+				result.Err = model.NewAppError("SqlPostStore.GetMaxPostSize", "store.sql_post.query_max_post_size.error", nil, err.Error(), http.StatusInternalServerError)
+			} else {
+				// Assume a worst-case representation of four bytes per rune.
+				maxPostSize = maxPostSizeBytes / 4
+
+				// To maintain backwards compatibility, don't yield a maximum post
+				// size smaller than the previous limit, even though it wasn't
+				// actually possible to store 4000 runes in all cases.
+				if maxPostSize < model.POST_MESSAGE_MAX_RUNES_V1 {
+					maxPostSize = model.POST_MESSAGE_MAX_RUNES_V1
+				}
+			}
+		} else {
+			l4g.Warn(utils.T("store.sql_post.query_max_post_size.unrecognized_driver"))
+		}
+
+		maxPostSizeCached = &maxPostSize
+		result.Data = maxPostSize
```

Comment:
If I add some calls to the test like this...

```go
func testMaxPostSize(t *testing.T, ss store.Store) {
	ss.Post().GetMaxPostSize(false)
	ss.Post().GetMaxPostSize(false)
	ss.Post().GetMaxPostSize(false)
	ss.Post().GetMaxPostSize(false)
	assert.EqualValues(t, model.POST_MESSAGE_MAX_RUNES_V2, (<-ss.Post().GetMaxPostSize(false)).Data.(int32))
	assert.EqualValues(t, model.POST_MESSAGE_MAX_RUNES_V2, (<-ss.Post().GetMaxPostSize(true)).Data.(int32))
}
```

it exposes your data race:

```
=== RUN   TestPostStore/MySQL/MaxPostSize
==================
WARNING: DATA RACE
Write at 0x00000218b448 by goroutine 96:
  github.com/mattermost/mattermost-server/store/sqlstore.SqlPostStore.GetMaxPostSize.func1()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/sqlstore/post_store.go:1280 +0x1e7
  github.com/mattermost/mattermost-server/store.Do.func1()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/store.go:25 +0x91

Previous write at 0x00000218b448 by goroutine 94:
  github.com/mattermost/mattermost-server/store/sqlstore.SqlPostStore.GetMaxPostSize.func1()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/sqlstore/post_store.go:1280 +0x1e7
  github.com/mattermost/mattermost-server/store.Do.func1()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/store.go:25 +0x91

Goroutine 96 (running) created at:
  github.com/mattermost/mattermost-server/store.Do()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/store.go:23 +0x74
  github.com/mattermost/mattermost-server/store/sqlstore.SqlPostStore.GetMaxPostSize()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/sqlstore/post_store.go:1233 +0xd4
  github.com/mattermost/mattermost-server/store/sqlstore.(*SqlPostStore).GetMaxPostSize()
      <autogenerated>:1 +0x8c
  github.com/mattermost/mattermost-server/store/storetest.testMaxPostSize()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/storetest/post_store.go:1791 +0xe1
  github.com/mattermost/mattermost-server/store/storetest.TestPostStore.func27()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/storetest/post_store.go:46 +0x54
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 94 (finished) created at:
  github.com/mattermost/mattermost-server/store.Do()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/store.go:23 +0x74
  github.com/mattermost/mattermost-server/store/sqlstore.SqlPostStore.GetMaxPostSize()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/sqlstore/post_store.go:1233 +0xd4
  github.com/mattermost/mattermost-server/store/sqlstore.(*SqlPostStore).GetMaxPostSize()
      <autogenerated>:1 +0x8c
  github.com/mattermost/mattermost-server/store/storetest.testMaxPostSize()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/storetest/post_store.go:1789 +0x75
  github.com/mattermost/mattermost-server/store/storetest.TestPostStore.func27()
      /Users/chris/go/src/github.com/mattermost/mattermost-server/store/storetest/post_store.go:46 +0x54
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
==================
```

This looks like a job for `sync.Once` if you remove the `allowFromCache` parameter.